librz/bin/format/elf/elf.c

```diff
@@ -156,133 +156,63 @@ static void init_shdr_sdb(ELFOBJ *bin) {
-		!Elf_(rz_bin_elf_read_word)(bin, &offset, &section->sh_link) ||
-		!Elf_(rz_bin_elf_read_word)(bin, &offset, &section->sh_info) ||
-		!Elf_(rz_bin_elf_read_word_xword)(bin, &offset, &section->sh_addralign) ||
-		!Elf_(rz_bin_elf_read_word_xword)(bin, &offset, &section->sh_entsize)) {
-		return false;
-	}
-
-	return true;
-}
-
-static bool read_shdr_entry(ELFOBJ *bin, size_t shdr_entry_index) {
-	ut64 offset = bin->ehdr.e_shoff + shdr_entry_index * sizeof(Elf_(Shdr));
-
-	if (!set_shdr_entry(bin, bin->shdr + shdr_entry_index, offset)) {
-		return false;
-	}
-
-	return true;
-}
-
-static bool read_shdr(ELFOBJ *bin) {
-	for (size_t i = 0; i < bin->ehdr.e_shnum; i++) {
-		if (!read_shdr_entry(bin, i)) {
-			return false;
-		}
-	}
-
-	return true;
-}
-
-static bool init_shdr_header(ELFOBJ *bin) {
-	bin->shdr = RZ_NEWS0(Elf_(Shdr), bin->ehdr.e_shnum);
-	if (!bin->shdr) {
-		perror("malloc (shdr)");
-		return false;
-	}
-
-	if (!read_shdr(bin)) {
-		return false;
-	}
-
-	return true;
-}
-
-static bool check_shdr_size(ELFOBJ *bin) {
-	ut32 shdr_size;
-
-	if (!UT32_MUL(&shdr_size, bin->ehdr.e_shnum, sizeof(Elf_(Shdr)))) {
-		return false;
-	}
-
-	if (!shdr_size || bin->ehdr.e_shoff + shdr_size > bin->size) {
-		return false;
-	}
-
-	return true;
-}
-
-static bool rz_bin_elf_init_shdr(ELFOBJ *bin) {
-	if (!bin->ehdr.e_shnum || !check_shdr_size(bin)) {
+static bool rz_bin_elf_init_shdr(ELFOBJ *bin, RzVector *sections) {
+	bin->sections = Elf_(rz_bin_elf_convert_sections)(bin, sections);
+	if (!bin->sections) {
 		return false;
 	}
 
 	init_shdr_sdb(bin);
 
-	return init_shdr_header(bin);
-}
-
-static void init_shstrtab_sdb(ELFOBJ *bin) {
-	sdb_num_set(bin->kv, "elf_shstrtab.offset", bin->shstrtab_section->sh_offset, 0);
-	sdb_num_set(bin->kv, "elf_shstrtab.size", bin->shstrtab_section->sh_size, 0);
-}
-
-static bool check_shstrtab(ELFOBJ *bin) {
-	return bin->shstrtab_section->sh_offset + bin->shstrtab_section->sh_size < bin->size;
+	return true;
 }
 
-static bool check_shstrtab_index(ELFOBJ *bin, Elf_(Half) shstrtab_index) {
-	return shstrtab_index != SHN_UNDEF && Elf_(rz_bin_elf_is_sh_index_valid)(bin, shstrtab_index);
+static void init_shstrtab_sdb(ELFOBJ *bin, ut64 offset, ut64 size) {
+	sdb_num_set(bin->kv, "elf_shstrtab.offset", offset, 0);
+	sdb_num_set(bin->kv, "elf_shstrtab.size", size, 0);
 }
 
-static bool set_shstrtab(ELFOBJ *bin) {
-	bin->shstrtab = calloc(1, bin->shstrtab_size + 1);
+static bool set_shstrtab(ELFOBJ *bin, Elf_(Off) offset) {
+	bin->shstrtab = RZ_NEWS(char, bin->shstrtab_size + 1);
 	if (!bin->shstrtab) {
-		perror("malloc");
 		return false;
 	}
 
-	int res = rz_buf_read_at(bin->b, bin->shstrtab_section->sh_offset, (ut8 *)bin->shstrtab, bin->shstrtab_section->sh_size);
+	int res = rz_buf_read_at(bin->b, offset, (ut8 *)bin->shstrtab, bin->shstrtab_size);
 	if (res < 0) {
-		bprintf("read (shstrtab) at 0x%" PFMT64x "\n", (ut64)bin->shstrtab_section->sh_offset);
+		bprintf("read (shstrtab) at 0x%" PFMT64x "\n", (ut64)offset);
```

Comment:
what is bprintf? and since this is a fail, please use RZ_LOG_ERR here

Reply:
It was a think I wanted to remove. It is a macro that printf if bin->verbose == true